pkg/runtime/profiler.go

```diff
@@ -0,0 +1,135 @@
+package runtime
+
+import (
+	"fmt"
+	"os"
+	goruntime "runtime"
+	"runtime/pprof"
+	"runtime/trace"
+)
+
+// ProfilerConfig defines the configuration of a Profiler
+type ProfilerConfig struct {
+	CPUProfile         bool
+	CPUProfileFileName string
+	MemProfile         bool
+	MemProfileRate     int
+	MemProfileFileName string
+	Trace              bool
+	TraceFileName      string
+}
+
+// Profiler defines the methods to control execution profiling
+type Profiler interface {
+	// Start initiates the tracing. If already active, has no effect
+	Start() error
+	// Stop terminates the tracing. If not started, has no effect
+	Stop() error
+}
+
+// profiler maintains the configuration state of the profiler
+type profiler struct {
+	ProfilerConfig
+	active         bool
```

Comment:
I'm not able to find any place where we set `active` to `true`, is this intended?

If possible I'd avoid capturing state in boolean flags, as in my experience they are a very common source of bugs. Can we avoid having this by stopping the profilers unconditionally? Users should not call Stop without having called Start in the first place, and I think stopping profilers is idempotent either way (at least for the CPU):

> StopCPUProfile stops the current CPU profile, **if any**. StopCPUProfile only returns after all the writes for the profile have completed.

Reply:
Good catch, it should have been used to avoid starting the profiling more than once and making the Stop a no-op if it was not active. 

As for not using a boolean, what would you then suggest to achieve the goals above :point_up:

Reply:
Interesting, I did not think about the double-start problem! My first hunch would be to just ignore this and trust the caller to not call this more than once (like the underlying profiler does).

The object-oriented way to avoid this kind of behavior without implicit state is having a `Start` method that returns a started profiler (that would be a different object), which can be `Stop`ped. Then, it is fine for users to call `Start` as many times as they want.

That being said, that pattern may an overkill for something simple as this. If we really have to check for this, the bool is okay.

Reply:
> My first hunch would be to just ignore this and trust the caller to not call this more than once (like the underlying profiler does).

I  agree that this is probably the best solution. And the fact we are not using this field makes it the "default" (even if unintended) Implementation.

> The object-oriented way to avoid this kind of behavior without an implicit state is having a Start method that returns a started profiler (that would be a different object), which can be Stopped. Then, it is fine for users to call Start as many times as they want.

That makes sense. Maybe we can change the `Profiler` method to `StartProfiling(config)` and return a simple interface with a `Stop` (or reuse `io.Closer` which already have the method `Close`)